feat: Segment PDF text into logical blocks (Experience, Education, etc.)

saviobatista opened the pull request:
### 📋 **Overview**
This PR closes #6 and implements a robust PDF text segmentation system that automatically extracts and categorizes résumé content into logical sections (Experience, Education, Skills, etc.). This feature forms the foundation for the AI-powered résumé tailoring system described in the project roadmap.

### ✨ **What This PR Adds**

#### **Core Functionality**
- **`segmentBlocks()` function**: Main segmentation engine that processes extracted PDF text
- **Section Detection**: Intelligent pattern matching for Experience, Education, and Skills sections
- **Content Processing**: Smart merging of bullet points and related content
- **Skills Extraction**: Automatic detection of technical skills from experience and education content
- **Confidence Scoring**: Quality assessment algorithm for segmentation results

#### **New Files Added**
- `lib/parser/segmentBlocks.ts` - Main segmentation implementation (355 lines)
- `lib/parser/config.ts` - Configuration for patterns and constants (90 lines)
- `lib/parser/types.ts` - TypeScript interfaces for internal processing (57 lines)
- `tests/parser/segmentBlocks.test.ts` - Unit tests (188 lines)
- `tests/parser/integration.test.ts` - Integration tests (109 lines)

#### **Enhanced Files**
- `lib/types.ts` - Added public interfaces and error classes (38 lines added)

### 🏗️ **Architecture & Design**

#### **Type System**
```typescript
interface ResumeBlock {
  content: string;
  confidence: number;
  startLine: number;
  endLine: number;
}

interface SegmentationResult {
  blocks: {
    experience?: string;
    education?: string;
    skills?: string;
    [key: string]: string | undefined;
  };
  warnings: string[];
  metadata: {
    totalBlocks: number;
    confidence: number;
    detectedSections: string[];
  };
}
```

#### **Configuration-Driven**
- **Section Patterns**: Configurable regex patterns for different résumé formats
- **Job Indicators**: Smart detection of job title and company patterns
- **Education Indicators**: Recognition of degree and university patterns
- **Skill Patterns**: Comprehensive technical skills detection

#### **Processing Pipeline**
1. **Text Normalization** → Clean and prepare input text
2. **Section Detection** → Identify section boundaries using patterns
3. **Content Processing** → Merge and structure section content
4. **Skills Extraction** → Extract technical skills from experience/education
5. **Result Building** → Generate structured output with metadata

### 🧪 **Testing & Quality**

#### **Test Coverage**
- **16 tests passing** with comprehensive coverage
- **Unit Tests**: Individual function testing with edge cases
- **Integration Tests**: End-to-end pipeline verification
- **Real Data**: Tests use actual résumé sample data

#### **Code Quality**
- ✅ **ESLint**: No warnings or errors
- ✅ **TypeScript**: Compilation successful
- ✅ **Coverage**: All critical paths tested
- ✅ **Error Handling**: Comprehensive error scenarios covered

### 🔧 **Technical Implementation Details**

#### **Smart Content Merging**
```typescript
// Merges bullet points with descriptions
function mergeBulletPoints(content: string): string {
  // Detects job/education patterns and merges related content
  // Handles various bullet point formats (-, •, etc.)
}
```

#### **Confidence Calculation**
```typescript
// Sophisticated scoring based on:
// - Section detection accuracy
// - Content length and quality
// - Pattern matching success
// - Overall structure integrity
```

#### **Error Handling**
```typescript
export class SegmentationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SegmentationError";
  }
}
```

### 📊 **Performance & Scalability**

- **Efficient Processing**: Single-pass text analysis with O(n) complexity
- **Memory Optimized**: Minimal memory footprint during processing
- **Configurable Limits**: Adjustable thresholds for different use cases
- **Extensible**: Easy to add new section types and patterns

### 🔗 **Integration Points**

#### **With Existing Code**
- **Seamless Integration**: Works with existing `extractText` function
- **Type Compatibility**: Full TypeScript support throughout
- **Error Propagation**: Proper error handling in the pipeline

#### **Future Extensions**
- **AI Integration**: Ready for AI-powered content analysis
- **Template System**: Foundation for LaTeX template generation
- **Custom Patterns**: User-configurable section detection

### 🎯 **Use Cases & Examples**

#### **Input Example**
```
Java Programming Experience
- Senior Java Developer – FinTech Solutions Inc. (2021–2023)
– Led microservices architecture handling $50M monthly transactions
```

#### **Output Structure**
```typescript
{
  blocks: {
    experience: "Senior Java Developer – FinTech Solutions Inc. (2021–2023)\nLed microservices architecture...",
    skills: "Java, microservices, Spring Boot, RESTful APIs, AWS"
  },
  metadata: {
    totalBlocks: 2,
    confidence: 0.85,
    detectedSections: ["experience", "skills"]
  }
}
```

### 🚀 **Impact & Benefits**

#### **For Users**
- **Automatic Parsing**: No manual résumé section identification needed
- **Structured Data**: Clean, organized content for AI processing
- **Quality Assurance**: Confidence scoring indicates parsing reliability

#### **For Developers**
- **Maintainable Code**: Clean architecture with clear separation of concerns
- **Extensible Design**: Easy to add new features and patterns
- **Comprehensive Testing**: Robust test suite ensures reliability

### 📋 **Testing Instructions**

1. **Run Tests**: `npm test` (all 16 tests should pass)
2. **Lint Check**: `npm run lint` (should be clean)
3. **Type Check**: `npx tsc --noEmit` (should compile successfully)
4. **Integration Test**: Verify PDF extraction + segmentation pipeline

### 🔮 **Future Roadmap Integration**

This feature directly supports the planned roadmap items:
- ✅ **Resume Parser** - Core segmentation functionality
- 🔄 **AI Prompt Generator** - Structured data for AI processing
- 🔄 **Tailored Experience Selector** - Categorized content for customization
-  **LaTeX PDF Renderer** - Structured sections for template generation

### **Breaking Changes**
- **None** - This is a new feature with no impact on existing functionality

### 🧹 **Code Review Checklist**

- [x] **Functionality**: All features work as expected
- [x] **Testing**: Comprehensive test coverage
- [x] **Documentation**: Clear code comments and interfaces
- [x] **Error Handling**: Proper error scenarios covered
- [x] **Performance**: Efficient algorithms and data structures
- [x] **Type Safety**: Full TypeScript support
- [x] **Linting**: Code follows project standards

---

**Ready for Review** 🎯 | **Tests Passing** ✅ | **Production Ready** 🚀

saviobatista committed:
feat(types): add interfaces and custom error class for résumé segmentation

- Introduced ResumeBlock interface to represent individual résumé sections
- Added SegmentationResult interface for complete segmentation output
- Implemented SegmentationError class for handling segmentation failures

saviobatista committed:
feat(parser): add configuration and types for section detection and processing

- Introduced SECTION_PATTERNS, JOB_INDICATORS, EDUCATION_INDICATORS, and CONFIDENCE_CONSTANTS for improved section detection.
- Added interfaces for SectionContent, ProcessingBlocks, FinalBlocks, SectionDetectionResult, SkillsExtractionResult, WarningContext, and ResultBuildingContext to enhance type safety during processing.

saviobatista committed:
feat(parser): implement segmentBlocks function for résumé segmentation

- Added segmentBlocks function to detect and segment résumé sections (experience, education, skills).
- Implemented normalization of section labels, merging of bullet points, and extraction of skills from content.
- Introduced confidence scoring based on section detection quality and added warnings for missing sections.
- Created unit tests to validate functionality and ensure correct segmentation of various résumé formats.

saviobatista committed:
test(parser): add integration tests for PDF text extraction and segmentation

- Introduced integration tests for the extractText and segmentBlocks functions.
- Validated text extraction from both single-page and multi-page PDF fixtures.
- Ensured segmentation results meet expected structure and confidence levels.
- Added end-to-end test for processing sample résumé text to verify block detection.

saviobatista commented on lib/parser/segmentBlocks.ts:
```diff
@@ -0,0 +1,355 @@
+}
+
+export { SegmentationError };
+export default segmentBlocks;
```
This would raise a problem on typescript because you need to export types and have a default export at same time.

saviobatista commented on tests/parser/integration.test.ts:
```diff
@@ -0,0 +1,109 @@
+import path from "node:path";
+
+import extractText from "../../lib/parser/extractText";
+import segmentBlocks from "../../lib/parser/segmentBlocks";
```
Nope, typescript standard.

JoaoMiottiTec commented on lib/parser/segmentBlocks.ts:
```diff
@@ -0,0 +1,355 @@
+
+  return label.toLowerCase().replace(/\s+/g, "_");
+}
+
```
If you use a type to select the keys, I think it's better use a loop over SECTION_PATTERNS.  This way the future maintenance is requires adding a new key  config

saviobatista replied:
Nice catch, fixed.

JoaoMiottiTec commented on lib/parser/segmentBlocks.ts:
```diff
@@ -0,0 +1,355 @@
+      detectedSections.push(normalizedSection);
+    }
+  }
+}
```
If return early this func plus add a helper, i think it turn the function more legible

saviobatista replied:
Refactored

JoaoMiottiTec commented on lib/parser/segmentBlocks.ts:
```diff
@@ -0,0 +1,355 @@
+  const finalBlocks: FinalBlocks = {};
+
+  for (const [sectionKey, contentArray] of Object.entries(blocks)) {
+    if (sectionKey === "experience" || sectionKey === "education") {
```
I think it's best to define these keys using Set to clear the session because this set prevents duplicate values.

saviobatista replied:
Nice catch, refactored.

JoaoMiottiTec commented on lib/parser/segmentBlocks.ts:
```diff
@@ -0,0 +1,355 @@
+ */
+function generateWarnings(context: WarningContext): string[] {
+  const { finalBlocks, detectedSections } = context;
+  const warnings: string[] = [];
```
Set i think it's better again because isn't duplicate if this messages don't send one more time in the  warning

saviobatista replied:
Not here, this is defined by the business logic and the typed data is only the result holder, as we have specific validation in this method for processing and assigning the warning values, the logic itself avoid duplication of warnings

JoaoMiottiTec commented on tests/parser/integration.test.ts:
```diff
@@ -0,0 +1,109 @@
+import path from "node:path";
+
+import extractText from "../../lib/parser/extractText";
+import segmentBlocks from "../../lib/parser/segmentBlocks";
```
Alias path i think it's better to future maintenance in the architecture

saviobatista replied:
Done, also I've refactored all imports from lib and tests folders. It needed also an extra configuration on vite config.

saviobatista committed:
refactor(parser): streamline section normalization and content handling

- Simplified the normalization of section labels by using a loop over SECTION_PATTERNS.
- Introduced a helper function, createSectionContent, to encapsulate section content creation logic.
- Updated section processing to utilize the new configuration for sections with bullet points, enhancing clarity and maintainability.

saviobatista committed:
chore: update vitest configuration and refactor import paths

- Added vite-tsconfig-paths plugin to vitest configuration for improved module resolution.
- Refactored import paths in parser files and tests to use absolute paths for better clarity and maintainability.

saviobatista committed:
chore: clean up formatting in documentation and configuration files

- Added missing newlines in docker-compose.yml for consistency.
- Improved formatting in README.md for better readability of commands.
- Enhanced clarity in architecture.md and parser.md by adding spacing for better structure.
